docs/essentials/gc/unmanaged/implementingdispose.md

```diff
@@ -1 +1,442 @@
+
+* The [SafeFileHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeFileHandle.html), [SafeMemoryMappedFileHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeMemoryMappedFileHandle.html), and [SafePipeHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafePipeHandle.html) class, for files, memory mapped files, and pipes. 
+
+* The [SafeMemoryMappedViewHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeMemoryMappedViewHandle.html) class, for memory views.
+
+*The [SafeNCryptKeyHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeNCryptKeyHandle.html), [SafeNCryptProviderHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeNCryptProviderHandle.html), and [SafeNCryptSecretHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeNCryptSecretHandle.html) classes, for cryptography constructs. 
+
+* The [SafeRegistryHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeRegistryHandle.html) class, for registry keys. 
+
+* The [SafeWaitHandle](http://dotnet.github.io/api/Microsoft.Win32.SafeHandles.SafeWaitHandle.html) class, for wait handles. 
+
+## Using a safe handle to implement the dispose pattern for a base class
+
+The following example illustrates the dispose pattern for a base class, `DisposableStreamResource`, that uses a safe handle to encapsulate unmanaged resources. It defines a `DisposableResource` class that uses a `SafeFileHandle` to wrap a `Stream` object that represents an open file. The `DisposableResource` method also includes a single property, `Size`, that returns the total number of bytes in the file stream. 
+
+```csharp
+using Microsoft.Win32.SafeHandles;
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+public class DisposableStreamResource : IDisposable
+{
+   // Define constants.
+   protected const uint GENERIC_READ = 0x80000000;
+   protected const uint FILE_SHARE_READ = 0x00000001;
+   protected const uint OPEN_EXISTING = 3;
+   protected const uint FILE_ATTRIBUTE_NORMAL = 0x80;
+   protected IntPtr INVALID_HANDLE_VALUE = new IntPtr(-1);
+   private const int INVALID_FILE_SIZE = unchecked((int) 0xFFFFFFFF);
+
+   // Define Windows APIs.
+   [DllImport("kernel32.dll", EntryPoint = "CreateFileW", CharSet = CharSet.Unicode)]
+   protected static extern IntPtr CreateFile (
+                                  string lpFileName, uint dwDesiredAccess, 
+                                  uint dwShareMode, IntPtr lpSecurityAttributes, 
+                                  uint dwCreationDisposition, uint dwFlagsAndAttributes, 
+                                  IntPtr hTemplateFile);
+
+   [DllImport("kernel32.dll")]
+   private static extern int GetFileSize(SafeFileHandle hFile, out int lpFileSizeHigh);
```

Comment:
`GetFileSize()` is documented to return `DWORD`, i.e. `uint`. Shouldn't this declaration follow that? This would also avoid the awkwardness of defining `INVALID_FILE_SIZE` using an `unchecked` cast.